src/EventListener/TwoFactorLoginSuccessHandler.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\EventListener;
+
+use Sonata\UserBundle\GoogleAuthenticator\Helper;
+use Sonata\UserBundle\Model\User;
+use Sonata\UserBundle\Model\UserManagerInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+
+class TwoFactorLoginSuccessHandler implements AuthenticationSuccessHandlerInterface
+{
+    /**
+     * @var AuthorizationCheckerInterface
+     */
+    private $authorizationChecker;
+
+    /**
+     * @var EngineInterface
+     */
+    private $engine;
+
+    /**
+     * @var Helper
+     */
+    private $googleAuthenticator;
+
+    /**
+     * @var UserManagerInterface
+     */
+    private $userManager;
+
+    /**
+     * @var array
+     */
+    private $forcedForRoles;
+
+    /**
+     * @var array
+     */
+    private $ipWhiteList;
+
+    /**
+     * AuthenticationSuccessHandler constructor.
+     *
+     * @param array $forcedForRole Roles that need to force use 2FA authorization
+     * @param array $ipWhiteList   IP's that will skip 2FA authorization
+     */
+    public function __construct(
+        AuthorizationCheckerInterface $authorizationChecker,
+        EngineInterface $engine,
+        Helper $helper,
+        UserManagerInterface $userManager,
+        array $forcedForRole,
+        array $ipWhiteList
+    ) {
+        $this->authorizationChecker = $authorizationChecker;
+        $this->engine = $engine;
+        $this->googleAuthenticator = $helper;
+        $this->userManager = $userManager;
+        $this->forcedForRoles = $forcedForRole;
+        $this->ipWhiteList = $ipWhiteList;
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token)
```

Comment:
<s>Please add a PHPDoc with the return type unless we can't use return types in the method signature</s>

Please add the return type to the signature

Reply:
@core23 t what method?

Reply:
@core23 add to both